// @nightwatch/mobile-helper/dist/commands/android/dotcommands.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AndroidDotCommand = void 0;
const ansi_colors_1 = __importDefault(require("ansi-colors"));
const dotenv = __importStar(require("dotenv"));
const path_1 = __importDefault(require("path"));
const constants_1 = require("../../constants");
const logger_1 = __importDefault(require("../../logger"));
const utils_1 = require("../../utils");
const common_1 = require("./subcommands/common");
const common_2 = require("./utils/common");
const sdk_1 = require("./utils/sdk");
class AndroidDotCommand {
    dotcmd;
    args;
    sdkRoot;
    rootDir;
    platform;
    androidHomeInGlobalEnv;
    constructor(dotcmd, argv, rootDir = process.cwd()) {
        this.dotcmd = dotcmd;
        this.args = argv.slice(1);
        this.sdkRoot = '';
        this.rootDir = rootDir;
        this.platform = (0, utils_1.getPlatformName)();
        this.androidHomeInGlobalEnv = false;
    }
    async run() {
        if (!constants_1.ANDROID_DOTCOMMANDS.includes(this.dotcmd)) {
            logger_1.default.log(ansi_colors_1.default.red(`Unknown dot command passed: ${this.dotcmd}\n`));
            logger_1.default.log('Run Android SDK command line tools using the following command:');
            logger_1.default.log(ansi_colors_1.default.cyan('npx @nightwatch/mobile-helper <DOTCMD> [options|args]\n'));
            logger_1.default.log(`Available Dot Commands: ${ansi_colors_1.default.magenta(constants_1.ANDROID_DOTCOMMANDS.join(', '))}`);
            logger_1.default.log(`(Example command: ${ansi_colors_1.default.gray('npx @nightwatch/mobile-helper android.emulator @nightwatch-android-11')})\n`);
            return false;
        }
        const javaInstalled = (0, common_2.checkJavaInstallation)(this.rootDir);
        if (!javaInstalled) {
            return false;
        }
        this.loadEnvFromDotEnv();
        const sdkRootEnv = (0, common_2.getSdkRootFromEnv)(this.rootDir, this.androidHomeInGlobalEnv);
        if (!sdkRootEnv) {
            logger_1.default.log(`Run: ${ansi_colors_1.default.cyan('npx @nightwatch/mobile-helper android --standalone')} to fix this issue.`);
            logger_1.default.log(`(Remove the ${ansi_colors_1.default.gray('--standalone')} flag from the above command if using the tool for testing.)\n`);
            return false;
        }
        this.sdkRoot = sdkRootEnv;
        const binaryName = this.dotcmd.split('.')[1];
        const binaryLocation = (0, common_2.getBinaryLocation)(this.sdkRoot, this.platform, binaryName, true);
        if (!binaryLocation) {
            (0, common_1.showMissingBinaryHelp)(binaryName);
            return false;
        }
        return (0, sdk_1.spawnCommandSync)(binaryLocation, binaryName, this.platform, this.args);
    }
    loadEnvFromDotEnv() {
        this.androidHomeInGlobalEnv = 'ANDROID_HOME' in process.env;
        dotenv.config({ path: path_1.default.join(this.rootDir, '.env') });
    }
}
exports.AndroidDotCommand = AndroidDotCommand;
