// @nightwatch/mobile-helper/dist/commands/android/subcommands/list/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.list = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const logger_1 = __importDefault(require("../../../../logger"));
const common_1 = require("../common");
const avd_1 = require("./avd");
const device_1 = require("./device");
async function list(options, sdkRoot, platform) {
    const optionsVerified = (0, common_1.verifyOptions)('list', options);
    if (!optionsVerified) {
        return false;
    }
    let subcommandFlag = optionsVerified.subcommandFlag;
    if (subcommandFlag === '') {
        subcommandFlag = await promptForFlag();
    }
    if (subcommandFlag === 'avd') {
        return await (0, avd_1.listInstalledAVDs)(sdkRoot, platform);
    }
    else if (subcommandFlag === 'device') {
        return await (0, device_1.listConnectedDevices)(sdkRoot, platform);
    }
    return false;
}
exports.list = list;
async function promptForFlag() {
    const flagAnswer = await inquirer_1.default.prompt({
        type: 'list',
        name: 'flag',
        message: 'Select what do you want to list:',
        choices: ['Connected devices', 'Installed AVDs']
    });
    logger_1.default.log();
    const flag = flagAnswer.flag;
    if (flag === 'Connected devices') {
        return 'device';
    }
    return 'avd';
}
