// @nightwatch/mobile-helper/dist/commands/android/subcommands/install/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.install = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const logger_1 = __importDefault(require("../../../../logger"));
const common_1 = require("../common");
const app_1 = require("./app");
const avd_1 = require("./avd");
const system_image_1 = require("./system-image");
async function install(options, sdkRoot, platform) {
    const optionsVerified = (0, common_1.verifyOptions)('install', options);
    if (!optionsVerified) {
        return false;
    }
    let subcommandFlag = optionsVerified.subcommandFlag;
    if (subcommandFlag === '') {
        subcommandFlag = await promptForFlag();
    }
    if (subcommandFlag === 'app') {
        return await (0, app_1.installApp)(options, sdkRoot, platform);
    }
    else if (subcommandFlag === 'system-image') {
        return await (0, system_image_1.installSystemImage)(sdkRoot, platform);
    }
    else if (subcommandFlag === 'avd') {
        return await (0, avd_1.createAvd)(sdkRoot, platform);
    }
    return false;
}
exports.install = install;
async function promptForFlag() {
    const flagAnswer = await inquirer_1.default.prompt({
        type: 'list',
        name: 'flag',
        message: 'Select what do you want to install:',
        choices: [
            { name: 'Android app (APK)', value: 'app' },
            { name: 'Android Virtual Device (AVD)', value: 'avd' },
            { name: 'System image', value: 'system-image' }
        ]
    });
    logger_1.default.log();
    return flagAnswer.flag;
}
